src/VecSim/algorithms/hnsw/hnsw_multi.h

```diff
@@ -212,6 +218,7 @@ HNSWIndex_Multi<DataType, DistType>::newBatchIterator(const void *queryBlob,
 template <typename DataType, typename DistType>
 std::vector<idType> HNSWIndex_Multi<DataType, DistType>::markDelete(labelType label) {
     std::vector<idType> idsToDelete;
+    std::shared_lock<std::shared_mutex> index_data_lock(this->index_data_guard_);
```

Comment:
right.
You're also right about holding the index data guard for shared ownership when replacing the entry point, it's something to consider. I believe we're OK since `markDelete` is called from the main thread only, and the other place were we change the entry point is upon inserting a new element. and there we're holding the index data guard for *exclusive lock*